cgrok/tree.c

```diff
@@ -0,0 +1,275 @@
+  TCTREE *tree = malloc(sizeof(TCTREE));
+  if (!tree) {
+    return NULL;
+  }
+  tree->dict = hb_dict_new(cp, tcfree);
+  tree->iter = NULL;
+  return tree; 
+}
+
+// Create an iterator to walk keys in ascending order
+// TokyoCabinet only allows one iterator per tree,
+// so we do the same.
+void tctreeiterinit(TCTREE *tree) {
+  if (tree->iter != NULL) {
+    dict_itor_free(tree->iter);
+  }
+  tree->iter = dict_itor_new(tree->dict);
+  dict_itor_first(tree->iter);
+}
+
+// Get the next key from the iterator
+const void *tctreeiternext(TCTREE *tree, int *sp) {
+  void *key = dict_itor_key(tree->iter);
+  dict_itor_next(tree->iter);
+  *sp = 0;
+  if (key) {
+    *sp = *(uint32_t*)key;
+    return key+4;
+  }
+  return NULL;
+}
+
+// Pack a key or value: uint32_t + body + NULL
+// Callers might use integers, structs, etc. for 
+// keys and values, so we can't depend on null-termination
+// to find the lenght of a key.
+void *tcpack(const void *value, uint32_t size) {
+  char *buf = malloc(size + 5);
+  if (!buf) {
+    return NULL;
+  }
+  memcpy(buf, &size, 4);
+  memcpy(buf+4, value, size);
+  buf[size+4] = 0;
+  return buf;
+}
+
+// Insert a key-value pair. If the key already exists the value will be overwritten
+void tctreeput(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    
+  }
+  void *val = tcpack(vbuf, vsiz);
+  if (!val) {
+
+  }
+  bool inserted; 
+  void ** valPtr = dict_insert(tree->dict, key, &inserted);
+  if (!inserted) {
+    free(key);
+    free(*valPtr);
+  }
+  *valPtr = val;
+}
+
+// Insert a key-value pair. If the key already exists return false and keep the original value
+bool tctreeputkeep(TCTREE *tree, const void *kbuf, int ksiz, const void *vbuf, int vsiz) {
+  void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    fprintf(stderr, "Failed to malloc tree key\n");
+    return;
+  } 
+  void *val = tcpack(vbuf, vsiz); 
+  if (!val) {
+    fprintf(stderr, "Failed to malloc tree value\n");
+    return;
+  }
+  bool inserted;
+  void **valPtr = dict_insert(tree->dict, key, &inserted);
+  if (inserted) {
+    *valPtr = val;
+  } else {
+    free(key);
+    free(val); 
+  }
+  return inserted;
+}
+
+// Get the value for the given key, or NULL if the key is not in the tree
+const void *tctreeget(TCTREE *tree, const void *kbuf, int ksiz, int *sp) {
+  const void *key = tcpack(kbuf, ksiz);
+  if (!key) {
+    fprintf(stderr, "Failed to malloc tree key\n");
+    return;
+  }
+  void *value = dict_search(tree->dict, key);
+  *sp = 0;
+  free(key);
+  if (value) {
+    *sp = *(uint32_t*)value;
+    return value+4;
+  }
+  return NULL;
+}
+
+// Remove all elements from the tree
+void tctreeclear(TCTREE *tree) {
+  dict_clear(tree->dict); 
+}
+
+// Free the tree and associated iterator
+void tctreedel(TCTREE *tree) {
+  dict_free(tree->dict);
+  if (tree->iter) {
+    dict_itor_free(tree->iter);
+  }
+  free(tree);
+}
+
+// A simple, singly linked list
```

Comment:
From my quick read of the TC spec `TCLIST` is an array list rather than a linked list. Does this change any performance assumptions of grok's use of `TCLIST`?